src/Adapter/Product/Provider/StockAvailableProvider.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Provider;
+
+use PrestaShop\PrestaShop\Adapter\AbstractObjectModelProvider;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductStockException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use StockAvailable;
+
+class StockAvailableProvider extends AbstractObjectModelProvider
+{
+    /**
+     * @param ProductId $productId
+     *
+     * @return StockAvailable
+     *
+     * @throws CoreException
+     */
+    public function getOrCreate(ProductId $productId): StockAvailable
```

Comment:
This could bring some discussions, but
personally I would avoid method which defines multiple repsonsibility in its name.
Idk if it is necessary, e.g. if its used in single place only, maybe this method is not needed at all ( you could check for ge/create in that specific place only)? Or maybe we could use something like `provide` for this case?

Reply:
You're right I split both methods and they are used by the ProductStockUpdater